### ISSUE TITLE
USB keyboard support in systemd-boot for ultrafast boot mode

### ISSUE THREAD

**Issue by kakra** [systemd/systemd#10827]
**Is your feature request related to a problem? Please describe.**
When running EFI BIOS in Ultra-Fast boot mode (or sometimes called "Windows 8 boot mode"), systemd-boot doesn't see my USB keyboard. I can only wait the timeout and let it boot the default menu item. I can work around it by forcing the mainboard into entering the EFI setup which initializes the EFI keyboard driver (holding the power button on cold start, it turns off with a "beep", next power on, I have keyboard support). When I exit EFI without saving the configuration, it continues loading systemd-boot and I now have keyboard control.

**Describe the solution you'd like**
systemd-boot should initialize the keyboard driver or somehow implement its own. I'm not sure how this is supposed to work inside EFI. When I run an EFI shell instead, it also has a working keyboard so it should be possible.

**Describe alternatives you've considered**
With my mainboard, there's a somewhat ugly workaround: You can force-enable keyboard support by turning the system off, then press and hold the power button until it turns on and then off again with a "beep" sound. Now press the power button once and there's working keyboard support. The EFI settings document that in ultra fast boot mode, USB devices (including keyboard/mouse) won't be initialized. But it's the only way to have a very fast boot with UEFI GOP support resulting in a smooth, fast, and flicker-free boot experience. Otherwise the POST takes over 10 seconds and my kernel even wouldn't properly boot with it most of the times (I'm not sure why but that's a different issue). Such a long POST adds 50% overhead to my boot time.

**Comment by medhefgo** [comment]
Afaik, the UEFI is allowed to skip all (USB) initialization that's not needed for boot (not a device it's booting from). I would expect the windows boot loader to also not be able to receive input when booted that way. You probably will have to use grub or disable fastboot, cause I doubt we can force loading USB drivers.

**Comment by kakra** [comment]
Windows doesn't have this problem as far as I can tell. But there's currently no existing Windows installation in this system. The manual of the mainboards states that in ultrafast boot mode, I need to invoke "reboot to firmware" from the OS to enter UEFI setup properly, this includes an example from the Windows 8 restore console.

Is there some API documentation how this all works or should work? I could imagine try working on an implementation.

**Comment by poettering** [comment]
I don't think we can do anything about it. Also, if ultra-fast boot is enabled, the whole point after all is not to initialize the USB stack to speed things up. If there was a way to forcibly initialize the USB stack, then people would be pissed because we defeat the whole effect the fast boot stuff is supposed to have.

I think people generally agree these days that the right approach to solve this is to accept that USB isn't initialized that early and instead make the fw setup more easily reachable from the post-boot UI. This is what windows does after all with its "reboot into firmware setup" feature. And systemd supports that too with `systemctl reboot --firmware-setup`. (And there's work in #10495 to open up more boot loader features in a similar way).

Hence, I don't see anything actionable here, fast boot works the way it works, we can't change it, and shouldn't fight it, but instead make the best of it, and I think we do with the reboot into firmware setup stuff"...

Hence closing... Or did I miss anything?

**Comment by kakra** [comment]
If I followed news correctly, you are working on some sort of fail-safe detection in sd-boot which could fall back to a previous kernel... That would be of great help already because you won't end up so easily stuck booting the wrong menu entry just because there's no keyboard working at the boot menu.

**Comment by poettering** [comment]
@kakra yupp, and that already landed, it's also a part of the puzzle, indeed.

**Comment by an4s911** [comment]
Is this solved. I have the same problem when trying to install ArcoLinux on an iMac. The keyboard doesn't work in the installation iso boot menu.

**Comment by kakra** [comment]
I don't think so, although the EFI of my new mainboard seems to actually enable USB even in ultra fast boot mode. But it also uses a lot longer time in firmware mode (5+ seconds vs 1.1 seconds). The only way would be disabling ultra fast boot mode but I'm not sure if this is possible with iMac.

From my experience, it was possible to get USB keyboards to work by entering the firmware setup, then exiting it again without any changes. This will prevent a reboot, and the keyboard driver is working.

I could also force my mainboard into normal boot mode with keyboard enabled when I powered the board off, and then powered it on by holding the power button down for 5 seconds. It will then power off again and power on in safe boot mode with ultra fast boot mode temporarily disabled.

On iMac, there may actually be a keycombo to hold down while powering on. Alternatively, you could modify the ISO boot menu to default to some other entry and wait for the timeout.

**Comment by medhefgo** [comment]
As already stated, fast boot modes generally don't load usb drivers (by design) to speed things up, but making it impossible for UEFI apps to get input from them. I am not aware of any API that lets us tell the firmware to initialize this on demand.

You could try booting into an efi shell and refind to see if input works for those. If that's the case then there's a chance for sd-boot.
Note that afaik grub (and probably windows boot manager) uses its own input stack and does not rely on UEFI provided APIs to get user input. That would be out of scope for sd-boot; you'd be better of using grub in that case.

**Comment by kakra** [comment]
From my memory:

Usually, efishells load the USB keyboard drivers and they work: Back when I tested it and systemd-boot won't detect the keyboard but had efishell as the default boot entry, the keyboard was working once efishell was running. If I then launched systemd-boot from EFI shell again, it would also see the keyboard. It's one of the work-arounds I was able to use.

So efishell somehow triggers loading of some drivers (even network as far as I remember), and efishell seems to be some generic binary I downloaded from somewhere (some trustworthy source, probably some board manufacturer)...
```
# md5sum /efi/shellx64.efi
01a8ee2759fa2170c065414489d9b823  /efi/shellx64.efi
```

I'm not sure where I got it from. It's been in place since several years. TianoCore also has a pre-compiled EFI shell:
* https://github.com/tianocore/edk2/blob/UDK2018/EdkShellBinPkg/FullShell/X64/Shell_Full.efi
* https://github.com/tianocore/edk2/blob/UDK2018/EdkShellBinPkg/MinimumShell/X64/Shell.efi
* https://github.com/tianocore/edk2/blob/UDK2018/ShellBinPkg/UefiShell/X64/Shell.efi

TL;DR It should be possible for sd-boot to somehow trigger loading the driver the same way EFI shells do it... Or maybe do they implement their own input layer? I'm not sure...

**Comment by an4s911** [comment]
> Alternatively, you could modify the ISO boot menu to default to some other entry and wait for the timeout.

Thanks, although I gave up on installing ArcoLinux (because I couldn't select the NVIDIA boot option) and instead installed vanilla Arch and setup the drivers, I would still like to know how I can change the boot order in an ISO boot menu. I tried opening the ISO with Archive manager and edited the files under the `loader` directory, but after I saved it, and then burned it to a USB, the USB wasn't recognized as bootable.

**Comment by kakra** [comment]
You probably need to extract the boot sector, too, and feed it back into the iso builder.

**Comment by medhefgo** [comment]
@kakra Can you try the `boot-reconnect` branch from https://github.com/medhefgo/systemd.git. There is a good chance it might make the firmware initialize things we need.

**Comment by kakra** [comment]
Meanwhile, I've switched to a new mainboard which supports the keyboard in fast boot mode (ultra fast doesn't seem to be a thing on modern mainboards which calibrate memory timings during boot).
